mentat/code_feature.py

```diff
@@ -157,67 +114,68 @@ def rel_path(self, cwd: Optional[Path] = None) -> str:
+    def get_code_message(self, standalone: bool = True) -> list[str]:
+        """
+        Gets this code features code message.
+        If standalone is true, will include the filename at top and extra newline at the end.
+        If feature contains entire file, will add inline diff annotations; otherwise, will append them to the end.
+        """
         session_context = SESSION_CONTEXT.get()
         code_file_manager = session_context.code_file_manager
         parser = session_context.config.parser
+        code_context = session_context.code_context
 
         code_message: list[str] = []
 
-        # We always want to give GPT posix paths
-        code_message_path = get_relative_path(self.path, session_context.cwd)
-        code_message.append(str(code_message_path.as_posix()))
-
-        if self.level in {CodeMessageLevel.CODE, CodeMessageLevel.INTERVAL}:
-            file_lines = code_file_manager.read_file(self.path)
-            for i, line in enumerate(file_lines):
-                if self.contains_line(i + 1):
-                    if parser.provide_line_numbers():
-                        code_message.append(
-                            f"{i + parser.line_number_starting_index()}:{line}"
-                        )
-                    else:
-                        code_message.append(f"{line}")
-        code_message.append("")
-
-        if self.diff is not None:
-            diff: str = get_diff_for_file(self.diff, self.path)
+        if standalone:
+            # We always want to give GPT posix paths
+            code_message_path = get_relative_path(self.path, session_context.cwd)
+            code_message.append(str(code_message_path.as_posix()))
+
+        # Get file lines
+        file_lines = code_file_manager.read_file(self.path)
+        for i, line in enumerate(file_lines):
+            if self.interval.contains(i + 1):
+                if parser.provide_line_numbers():
+                    code_message.append(
+                        f"{i + parser.line_number_starting_index()}:{line}"
+                    )
+                else:
+                    code_message.append(f"{line}")
+
+        if standalone:
+            code_message.append("")
+
+        if code_context.diff_context is not None:
+            diff = get_diff_for_file(code_context.diff_context.target, self.path)
             diff_annotations = parse_diff(diff)
-            if self.level == CodeMessageLevel.CODE:
+            if self.interval.whole_file():
                 code_message = annotate_file_message(code_message, diff_annotations)
             else:
                 for section in diff_annotations:
-                    code_message += section.message
+                    # TODO: Place diff_annotations inside interval where they belong
+                    if (
+                        section.start >= self.interval.start
+                        and section.start < self.interval.end
+                    ):
```

Comment:
This isn't actually an intersection check; this is checking to ensure that the diff annotation _starts_ within the interval (if it ends inside of the interval, we don't want to include it because a. it could be included in multiple intervals and b. the start of the diff annotation wouldn't make much sense because those line numbers aren't in this interval)